[Week7] Implement Authorization

JamieShin0201 opened the pull request:
# 해야 할 일
- [X] 회원 비밀번호 암호화
- [X] 오직 내 정보만 수정할 수 있도록 구현

### 추가적으로 할 일
- [ ] 전체적으로 리팩토링
- [X] 테스트코드 명확하게 작성
- [ ] 필요한 클래스에 JavaDoc 작성 

## 요구 사항
- [ ] 테스트 커버리지 100%를 달성해야 합니다.
- [X] 모든 API 테스트를 통과해야 합니다.
- [X] 모든 E2E 테스트를 통과해야 합니다.

JamieShin0201 committed:
feat: Spring Security 추가

JamieShin0201 committed:
feat: 기존 Interceptor 방식의 JWT 토큰 인증 방식을 Spring Security를 적용해 변경

JamieShin0201 committed:
feat: Password Encoder 추가 및 회원 비밀번호 암호화

JamieShin0201 committed:
feat: UserRegistrartionData, UserModificationData에 기본 생성자 추가

JamieShin0201 committed:
feat: 자신의 계정만 수정할 수 있도록 구현

JamieShin0201 commented:
# 2021/03/09 작업내용
1. 강의를 보며 Spring Security를 학습하고 과제에 적용하였습니다.

아직 Spring Security에 대한 학습이 부족하여ㅠ  잘 적용했는지 모르겠네요 ㅠ.ㅠ
계속 학습하면서 수정할 부분들 수정하겠습니다!

이번 한주도 잘 부탁 드립니다 ㅎ.ㅎ

hannut91 commented:
이번 주도 화이팅입니다 ㅎㅎ Spring Security가 상당히 복잡해서 처음부터 공부하다보면 엄청난 양에 압도당합니다 ㅎㅎ 제가 경험해봐서 허허.. 최소한의 기능만 구현해보면서 익힌다고 생각하시면 좋을 것 같습니다 ㅎㅎ

JamieShin0201 committed:
refactor: UserControllerTest를 BDD 형식으로 변경

JamieShin0201 committed:
fix: JwtAuthenticationFilter안의 Authoriztion, Beaer 를 의미있는 상수로 변경

JamieShin0201 committed:
fix: 접근 권한을 PreAuthorize에서 확인하도록 수정

JamieShin0201 commented:
# 2021/03/10 작업내용
1. UserControllerTest를 BDD 스타일로 변경하였습니다.
2. 회원 수정시 권한확인을 spring security가 제공하는 @PreAuthorize에서 처리하도록 변경하였습니다.

감사합니다 ^___^

hannut91 commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFilter.java:
```diff
@@ -18,6 +18,9 @@
 
     private final AuthenticationService authenticationService;
 
+    private static final String TOKEN_KEY = "Authorization";
```
`TOKEN_KEY`라는 이름이 조금 저는 이해하기 어려운 것 같아요. 인증 정보를 HTTP Header 를 통해 전달하고, `Authorization`속성을 이용하는 것인데 `TOKEN_KEY`라는 이름을 사용하는게 적절한가? 생각이 들었어요.
아니면 `인증 헤더`같은 의미가 들어있는 것은 어떨까요?

JamieShin0201 replied:
더 적절한 것 같습니다! AUTHENTICATION_HEADER 로 변경하였는데 괜찮을까요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -39,15 +38,12 @@ UserResultData create(@RequestBody @Valid UserRegistrationData registrationData)
 
     @PatchMapping("{id}")
-    @PreAuthorize("isAuthenticated()")
+    @PreAuthorize("isAuthenticated() and #id == authentication.principal")
```
파라미터 `id`변수와 `authentication`의 `principal`이 같은 경우를 보고 있는 거군요

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -5,153 +5,272 @@
+        class Context_with_invalid_access_token {
+            @Test
+            @DisplayName("상태코드 403 Forbidden 를 응답한다.")
+            void it_responds_status_code_401() throws Exception {
```
`it_responds_status_code_403`이 맞을 것 같습니다

JamieShin0201 replied:
제가 이전에 작성한 테스트 코드를 복사하고 필요한 부분만 수정하다가.. 실수를 했네요! 꼼꼼하지 못했습니다 ㅠ.ㅠ

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -5,153 +5,272 @@
-        mockMvc.perform(
-                patch("/users/100")
+        @Nested
+        @DisplayName("액세스 토큰이 유효 하지 않다면")
```
토큰이 유효하지만, 다른 사람의 계정 정보를 수정하는 경우의 테스트도 있을까요?

JamieShin0201 replied:
다른 사람의 계정정보에 수정요청을 보내는 경우 접근 권한이 없다는 예외가 발생해야 할 것 같습니다.
테스트 코드를 추가하였습니다.

JamieShin0201 committed:
fix: 상수명을 이해하기 쉽게 변경

TOKEN_KEY -> AUTHENTICATION_HEADER 로 변경

JamieShin0201 committed:
fix: UserControllerTest안의 테스트 함수 오타 수정

JamieShin0201 committed:
test: '존재하지 않는 회원 id가 주어질 경우' 회원 수정에 대한 테스트 코드 추가

JamieShin0201 committed:
feat: 관리자를 추가하고 관리자는 모든 회원을 수정할 수 있도록 변경

회원 도메인, JWT 토큰, 회원 인증 객체에 Role 추가

JamieShin0201 commented:
# 2021/03/11 작업내용
1. 리뷰를 확인하고 전반적으로 수정하였습니다.
2. 유저 도메인에 Role을 추가하고 관리자는 모든 회원을 수정할 수 있도록 변경하였습니다.

감사합니다 ㅎ.ㅎ

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -31,9 +34,11 @@
     private static final String INVALID_TOKEN = "eyJhbGciOiJIUzI1NiJ9." +
-            "eyJ1c2VySWQiOjF9.ZZ3CUl0jxeLGvQ1Js5nG2Ty5qGTlqai5ubDMXZOdaD0";
+            "eyJ1c2VySWQiOjEsInJvbGUiOiJST0xFX1VTRVIifQ." +
+            "SgLtYfTVUdvPF-gIP006U-_B7-wWMSUcJD3eoSOxHsE_INVALID";
```
길어서 복잡해지는군요 ㅎㅎ `VALID_TOKEN + "WRONG"` 처럼 표현하는 것도 좋을 것 같아요

hannut91 reviewed (commented):
Role을 추가해주셨군요 ㅎㅎ 더 나아가면 Entity로도 관리할 수 있을 것 같아요

JamieShin0201 committed:
유효하지 않은 토큰 값을 수정하라

유효하지 않은 토큰 값이 문자열로 되어있었습니다.
하지만, 불필요하게 너무 길고 한번에 파악하기 어려웠습니다.
따라서, 유효한 토큰을 가진 변수에 "WRONG" 이라는 문자열을 추가하도록 수정하였습니다.

JamieShin0201 committed:
Role을 엔터티로 변경하라

Role은 ROLE_USER, ROLE_ADMIN 두개를 가진 enum으로 되어있었습니다.
하지만, 추후에 Role이 더 많이질 것으로 예상되어 Role을 엔터티로 수정하였습니다.

JamieShin0201 commented:
# 2021/03/12 작업내용
1. 성규 동기님이 작성하신 커밋메시지를 참고하여 커밋메시지를 작성해 보았습니다.
2. Role을 엔터티로 변경하였습니다.

감사합니다 ㅎ.ㅎ

hannut91 commented:
> 유효하지 않은 토큰 값을 수정하라
> 
> 유효하지 않은 토큰 값이 문자열로 되어있었습니다.
> 하지만, 불필요하게 너무 길고 한번에 파악하기 어려웠습니다.
> 따라서, 유효한 토큰을 가진 변수에 "WRONG" 이라는 문자열을 추가하도록 수정하였습니다.

유효하지 않은 토큰 값을 수정하는 것으로는 이 작업의 진짜 의도를 파악하기 어렵습니다. 특히 `수정`이라는 말은 너무 많은 의미를 가지고 있습니다. 버그가 있어서 수정할 수도 있고, 개선의 의도로 수정할 수도 있고 새로운 기능으로 인해 수정할 수 있습니다. 따라서
`테스트를 위한 유효하지 않은 토큰 값을 단순화하라` 같이 진짜 의도를 드러내면 좋을 것 같아요

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -33,11 +40,17 @@ public User registerUser(UserRegistrationData registrationData) {
         User user = mapper.map(registrationData, User.class);
 
+        Role role = user.getRole();
+        Role foundRole = roleRepository.findByName(role.getName())
+                .orElseGet(() -> roleRepository.save(role));
```
우리 스터디에서 얘기했던 경우를 바로 작성해주시다니 ㅎㅎ 지금 이 라인에서 role로 검색해서 얻기만 할 것 같은데 새로운 롤을 생성하는 두 가지 일을 동시에 하고 있습니다 ㅎㅎ

JamieShin0201 replied:
그러네요.. ㅠㅠ 검색만 할 것 같은데 검색이 안되면 새로운 롤을 생성하는군요...

JamieShin0201 committed:
저장소에 주어진 Role이 없을 경우 예외를 발생시켜라

 Role이 존재하는지 확인하고 존재하지 않는다면 Role을 저장소에 추가하도록 되어있었습니다.

잘못된 Role이 입력되었을 경우에 원치 않는  Role이 추가되는 문제가 발생할 수 있습니다.
따라서, 주어진 Role이 저장소에 존재하지 않을 경우  예외를 발생시키도록 변경하였습니다.

JamieShin0201 committed:
'Role이 존재하지 않을 경우' 에 대한 HTTP 응답을 처리하라

회원을 등록할 때 Role이 존재하지 않는다면 예외를 발생시키도록 되어있습니다.
하지만, 발생한 예외를 클라이언트에게 응답해주는 코드가 작성되어 있지 않았습니다.
따라서, 'Role이 존재하지 않을 경우'에 대한 HTTP 응답을 처리하는 코드를 추가하였습니다.

JamieShin0201 committed:
사용하지 않는 코드를 삭제하라

ControllerAdvice와 UserAuthentication에 사용하지 않는 코드들이 있었습니다.
불필요한 코드라고 생각하여 삭제하였습니다.

JamieShin0201 committed:
UserAuthentication에 대한 테스트를 추가하라

UserAuthneticatoin에 대한 테스트코드가 없었습니다.
해당 객체가 원하는대로 만들어지는지 확인하기 위해 테스트 코드를 추가하였습니다.

hannut91 commented:
JWT를 통해 이 사용자가 어떤 권한을 가지고 있는지 노출될 수도 있을 것 같은데 상관없을까요?

JamieShin0201 commented:
> JWT를 통해 이 사용자가 어떤 권한을 가지고 있는지 노출될 수도 있을 것 같은데 상관없을까요?

JWT 토큰을 다른 사람이 알게되면 ADMIN 계정이 무엇인지 알수 있겠네요.. 미처 생각하지 못했습니다 ㅠ

hannut91 commented:
무엇을 학습하셨는지 너무 명확해서 좋네요 ㅎㅎ 아닛 그것보다 스터디를 두 개나 진행중이시다니... 대단하십니다 ㅎㅎ
다른 사람들이 내 코드를 이해하기 쉬울까? 라는 고민이 참 좋은 것 같습니다 ㅎㅎ 남이 보기 좋은 코드는 내가 보기에도 좋다라는 마음으로 
고민해보시면 좋을 것 같습니다 ㅎㅎ 고생하셨습니다!
